src/Compilers/CSharp/Portable/BoundTree/Statement.cs

```diff
@@ -310,20 +320,44 @@ partial class BoundStateMachineScope
         protected override OperationKind StatementKind => OperationKind.None;
     }
 
-    partial class BoundLocalDeclaration : IVariableDeclarationStatement, IVariable
+    partial class BoundLocalDeclaration : IVariableDeclarationStatement
     {
-        ImmutableArray<IVariable> IVariableDeclarationStatement.Variables => ImmutableArray.Create<IVariable>(this);
-
-        ILocalSymbol IVariable.Variable => this.LocalSymbol;
+        private static readonly ConditionalWeakTable<BoundLocalDeclaration, object> s_variablesMappings =
+            new ConditionalWeakTable<BoundLocalDeclaration, object>();
 
-        IExpression IVariable.InitialValue => this.InitializerOpt;
+        ImmutableArray<IVariable> IVariableDeclarationStatement.Variables
+        {
+            get
+            {
+                return (ImmutableArray<IVariable>) s_variablesMappings.GetValue(this, 
+                    declaration => ImmutableArray.Create<IVariable>(new VariableDeclaration(declaration.LocalSymbol, declaration.InitializerOpt, declaration.Syntax)));
+            }
+        }
 
         protected override OperationKind StatementKind => OperationKind.VariableDeclarationStatement;
     }
 
     partial class BoundMultipleLocalDeclarations : IVariableDeclarationStatement
     {
-        ImmutableArray<IVariable> IVariableDeclarationStatement.Variables => this.LocalDeclarations.As<IVariable>();
+        private static readonly ConditionalWeakTable<BoundMultipleLocalDeclarations, object> s_variablesMappings =
+            new ConditionalWeakTable<BoundMultipleLocalDeclarations, object>();
+
+        ImmutableArray<IVariable> IVariableDeclarationStatement.Variables
+        {
+            get
+            {
+                return (ImmutableArray<IVariable>) s_variablesMappings.GetValue(this,
+                    multipleDeclarations =>
+                    {
+                        var builder = ArrayBuilder<IVariable>.GetInstance(multipleDeclarations.LocalDeclarations.Length);
+                        foreach (var declaration in multipleDeclarations.LocalDeclarations)
+                        {
+                            builder.Add((IVariable)new VariableDeclaration(declaration.LocalSymbol, declaration.InitializerOpt, declaration.Syntax));
+                        }
+                        return builder.ToImmutableAndFree();
+                    });
```

Comment:
Perhaps add `BoundLocalDeclaration.ToVariable` helper method that could be used here and in `BoundLocalDeclaration.Variables`.